### PR TITLE
feat: Add auto-generated topic tags

### PR DESCRIPTION
**Pull request by oranges-eating**
http://82.156.7.146:7000/
usename:hsluoyz
password:123

### CODE REVIEW

**Comment by hsluoyz** [comment]
@oranges-eating remove merge commits

image

**Comment by hsluoyz** [comment]
@kocoler @Kininaru plz review.

File: service/tokenizer.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	"unicode/utf8"
+
+	"github.com/huichen/sego"
+)
+
+var segmenter sego.Segmenter
+var onceload sync.Once
+
+func Finalword(s string) []string {
+	Loaddic()
```
Comment (kocoler) [inline service/tokenizer.go]: This could be a init func.

File: service/tokenizer.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	var filterword, sortword, finalword []string
+	var sortnumber []int
+	wordnum := make(map[string]int)
+	for _, val := range word {
+		if utf8.RuneCountInString(val) > 1 {
+			filterword = append(filterword, val)
+		}
+	}
+	if len(filterword) == 0 {
+		return word
```
Comment (kocoler) [inline service/tokenizer.go]: If `filterword == 0` (means no filterword length logger than 1), but `len(word)` could be more than 5.

File: service/tokenizer.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	} else if len(filterword) < 5 {
+		return filterword
```
Comment (kocoler) [inline service/tokenizer.go]: `filterword` may contain duplicate words.

File: service/tokenizer.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		}
+		for s, v := range wordnum {
+			sortword = append(sortword, s)
+			sortnumber = append(sortnumber, v)
+		}
+		if len(sortword) < 5 {
+			return sortword
+		}
+		for i := 1; i < len(sortnumber); i++ {
+			if sortnumber[i] < sortnumber[i-1] {
```
Comment (kocoler) [inline service/tokenizer.go]: Too slow on a large page.
Make word and number in a struct and use `sort.Slice`.

**Review by kocoler — CHANGES_REQUESTED** [review]

File: Tag_test.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package main
```
Comment (kocoler) [inline Tag_test.go:15]: License.

**Comment by kocoler** [comment]
LGTM

**Review by keainye — APPROVED** [review]

File: Tag_test.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+var adapter *object.Adapter
+
+func TestTopicTag(t *testing.T) {
+	topics := []*object.Topic{}
+	adapter = object.NewAdapter(beego.AppConfig.String("driverName"), beego.AppConfig.String("dataSourceName"), beego.AppConfig.String("dbName"))
+	err := adapter.Engine.Table("topic").Find(&topics)
+	if err != nil {
+		t.Error("获取主题信息error")
```
Comment (hsluoyz) [inline Tag_test.go]: Use English & delete `t.Error` line

File: Tag_test.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+		panic(err)
+	}
+
+	for _, topic := range topics {
+		if len(topic.Tags) == 0 || topic.Tags == nil {
+			topic.Tags = service.Finalword(topic.Content)
+			_, err := adapter.Engine.Id(topic.Id).Update(topic)
+			if err != nil {
+				t.Error("更新主题tag error")
```
Comment (hsluoyz) [inline Tag_test.go]: Use English & delete `t.Error` line

File: service/tokenizer.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"sort"
+	"unicode/utf8"
+
+	"github.com/huichen/sego"
```
Comment (hsluoyz) [inline service/tokenizer.go:21]: go.mod update?

File: service/tokenizer.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+var segmenter sego.Segmenter
+
+type wordfreq struct {
+	word string
+	num  int
+}
+
+func init() {
+	segmenter.LoadDictionary("./dictionary/dictionary.txt")
```
Comment (hsluoyz) [inline service/tokenizer.go:32]: Is dictionary.txt in this repo?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: service/tokenizer.go (hunk 9 above)
Comment (oranges-eating) [reply service/tokenizer.go:32]: The dictionary used  in previous projects, and sego has also used them before

**Comment by hsluoyz** [comment]
@oranges-eating plz provide demo link for every PR.

File: service/tokenizer.go (hunk 8 above)
Comment (oranges-eating) [reply service/tokenizer.go:21]: Sego has been imported in the previous project, go.mod already has it

File: web/src/main/NewNodeTopicBox.js [hunk 10]
```diff
@@ -38,8 +38,10 @@ const ReactMarkdown = require("react-markdown");
 class NewNodeTopicBox extends React.Component {
   constructor(props) {
     super(props);
+    this.auto = React.createRef();
     this.state = {
       classes: props,
+      autotag: false,
```
Comment (hsluoyz) [inline web/src/main/NewNodeTopicBox.js]: autotag or autoTag ?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Review by oranges-eating — COMMENTED** [review]
> @oranges-eating plz provide demo link for every PR.

http://82.156.7.146:7000/

**Review by oranges-eating — COMMENTED** [review]
> @oranges-eating plz provide demo link for every PR.

http://82.156.7.146:7000/

**Pull request merged**